File: client/Classes/Battle/TouchLayer.cpp
//
//  TouchLayer.cpp
//  LuckyCat
//
//  Created by XiaoZhong Zhu on 12-8-23.
//  Copyright (c) 2012年 厦门雅基软件有限公司. All rights reserved.
//

#include "TouchLayer.h"

USING_NS_CC;

TouchLayer::TouchLayer():m_bTouchForbidden(false)
,m_bTouchAreaEnabled(false)
{
    m_TouchNodeList.clear();
}

TouchLayer::~TouchLayer()
{
    m_TouchNodeList.clear();
}

bool TouchLayer::init()
{
    return CCLayer::init();
}

void TouchLayer::onEnter()
{
    CCLayer::onEnter();
    
    this->setTouchEnabled(true);
}

void TouchLayer::onExit()
{
    CCLayer::onExit();
}

bool TouchLayer::ccTouchBegan(cocos2d::CCTouch *pTouch, cocos2d::CCEvent *pEvent)
{
    if ( !pTouch ) return false;
    
    if ( getIsTouchAreaEnabled() ) {
        CCPoint tPoint = this->convertTouchToNodeSpace(pTouch);
        if ( this->TouchAreaContainsPoint(tPoint) == false )
        {
            this->notificationTouchEvent(kLTouchEventOutsideTouchArea);
            
            return false;
        }
        else {
            this->notificationTouchEvent(kLTouchEventInsideTouchArea);
        }
    }
    else if ( getIsTouchForbidden() )
    {
        pBeginPoint = CCPointZero;
        return false;
    }
    
    pBeginPoint = this->convertTouchToNodeSpace(pTouch);
    
    return true;

}

void TouchLayer::ccTouchMoved(cocos2d::CCTouch *pTouch, cocos2d::CCEvent *pEvent)
{
    
}

void TouchLayer::ccTouchEnded(cocos2d::CCTouch *pTouch, cocos2d::CCEvent *pEvent)
{
    if ( !pTouch ) return;
    
    if ( pBeginPoint.x != 0 && pBeginPoint.y != 0 )
    {
        this->notificationTouchEvent(kLTouchEventSingleClick);
    }
    pBeginPoint = CCPointZero;
}

void TouchLayer::registerWithTouchDispatcher(void)
{
    CCDirector::sharedDirector()->getTouchDispatcher()->addTargetedDelegate(this, kCCMenuHandlerPriority , true);
}

bool TouchLayer::registerTouchNode(cocos2d::CCNode * pNode)
{
    if ( pNode )
    {
        std::vector<cocos2d::CCNode *>::iterator _iter = m_TouchNodeList.begin();
        for (; _iter != m_TouchNodeList.end(); _iter++) {
            CCNode *pN = *_iter;
            if ( pN == pNode )
            {
                return false;
            }
        }
        m_TouchNodeList.push_back(pNode);
        return true;
    }
    return false;
}

bool TouchLayer::removeTouchNode(cocos2d::CCNode * pNode)
{
    if ( pNode )
    {
        std::vector<cocos2d::CCNode *>::iterator _iter = m_TouchNodeList.begin();
        for (; _iter != m_TouchNodeList.end(); _iter++) {
            CCNode *pN = *_iter;
            if ( pN == pNode )
            {
                m_TouchNodeList.erase(_iter);
                return true;
            }
        }
    }
    return false;
}

bool TouchLayer::TouchAreaContainsPoint(cocos2d::CCPoint _point)
{
    std::vector<cocos2d::CCNode *>::iterator _iter = m_TouchNodeList.begin();
    for (; _iter != m_TouchNodeList.end(); _iter++) {
        CCNode *pN = *_iter;
        if ( pN )
        {
            CCRect tRect = getRect(pN);
            
            if ( CCRect::CCRectContainsPoint(tRect, _point) )
            {
                return true;
            }
        }
    }
    return false;
}

cocos2d::CCRect TouchLayer::getRect(cocos2d::CCNode * pNode)
{
    CCRect rc;
    rc.origin = pNode->getPosition();
    rc.size = pNode->getContentSize();
    
    if (pNode->isIgnoreAnchorPointForPosition() == false)
    {
        CCPoint tPoint = pNode->getAnchorPoint();
        
        rc.origin.x -= rc.size.width * tPoint.x;
        rc.origin.y -= rc.size.height * tPoint.y;
    }
    return rc;
}

File: server/src/browser/client/Classes/Event/TouchLayer.js
/**
 * Created with JetBrains WebStorm.
 * User: zhuxiaozhong
 * Date: 12-9-4
 * Time: 下午4:33
 * To change this template use File | Settings | File Templates.
 */
var lc = lc = lc || {};

/**
 * 这个类用于处理Touch事件,
 * 可扩展为手势处理
 */

/**
 * LTouchEvent 点击事件类型
 */
lc.kLTouchEventSingleClick              =   1;          //点击
lc.kLTouchEventDoubleClick              =   2;          //双击
lc.kLTouchEventSliding                  =   3;          //滑动
lc.kLTouchEventOutsideTouchArea         =   4;          //点击到点击区域之外
lc.kLTouchEventInsideTouchArea          =   5;          //点击到点击区域之内

lc.TouchLayer = cc.Layer.extend({
    m_bTouchForbidden : false,
    m_bTouchAreaEnabled : false,
    mBeginPoint : cc.PointZero(),
    m_TouchNodeList : [],
    pNode : null,
    ctor:function () {
        this._super();
    },
    init:function () {
        this.setTouchEnabled(true);
        return true;
    },
    registerWithTouchDispatcher:function () {
        cc.Director.getInstance().getTouchDispatcher().addTargetedDelegate(this,cc.MENU_HANDLER_PRIORITY + 10 , true);
    },
    onTouchBegan:function (touch, event) {
        cc.Assert(false, "Layer#onTouchBegan override me");
        if ( !touch ) return false;

        if ( this.getIsTouchAreaEnabled() )
        {
            var tPoint = touch._point;
            var _tPoint = this.convertTouchToNodeSpace(touch);

            cc.log("Orign x : " + tPoint.x + ", y: " + tPoint.y);
            cc.log("Conve x : " + _tPoint.x + ", y: " + _tPoint.y);
            if ( this.TouchAreaContainsPoint(tPoint) == false )
            {
                this.notificationTouchEvent(lc.kLTouchEventOutsideTouchArea);
            }
            else
            {
                this.notificationTouchEvent(lc.kLTouchEventInsideTouchArea);
            }
        }

        if ( this.getIsTouchForbidden() )
        {
            this.mBeginPoint.x = 0;
            this.mBeginPoint.y = 0;

            cc.log("Begin Touch Forbidden.");

            return false;
        }

        this.mBeginPoint.x = touch._point.x;
        this.mBeginPoint.y = touch._point.y;

        return true;
    },
    onTouchMoved:function (touch, event) {
    },
    onTouchEnded:function (touch, event) {
        if ( !touch ) return;

        if ( this.mBeginPoint.x != 0 && this.mBeginPoint.y != 0 )
        {
            this.notificationTouchEvent(lc.kLTouchEventSingleClick);
            cc.log("onTouchEnded Touch kLTouchEventSingleClick.");
        }

        this.mBeginPoint.x = 0;
        this.mBeginPoint.y = 0;
    },
    /**
     * 想要获得Touch事件的通知,必须重写这个函数
     */
    notificationTouchEvent:function (tLTouchEvent)
    {
        if ( tLTouchEvent == lc.kLTouchEventOutsideTouchArea )
        {
            cc.log("kLTouchEvent Outside TouchArea");
        }
        else if ( tLTouchEvent == lc.kLTouchEventInsideTouchArea )
        {
            cc.log("kLTouchEvent Inside TouchArea");
        }
        else if (tLTouchEvent == lc.kLTouchEventSingleClick)
        {
            cc.log("kLTouchEvent SingleClick");
        }
    },
    registerTouchNode:function (pNode)
    {
        if (pNode)
        {
            for (var i = 0; this.m_TouchNodeList && i < this.m_TouchNodeList.length ; i++)
            {
                var pN = this.m_TouchNodeList[i];
                if ( pN == pNode )
                {
                    return false;
                }
            }
            this.m_TouchNodeList.push(pNode);
            return true;
        }
        return false;
    },
    removeTouchNode:function (pNode)
    {
        if (pNode)
        {
            for (var i = 0; this.m_TouchNodeList && i < this.m_TouchNodeList.length ; i++)
            {
                var pN = this.m_TouchNodeList[i];
                if ( pN == pNode )
                {
                    this.m_TouchNodeList.splice(i,1);
                    return true;
                }
            }
        }
        return false;
    },
    TouchAreaContainsPoint:function (tPoint)
    {
        for (var i = 0; this.m_TouchNodeList && i < this.m_TouchNodeList.length; i++)
        {
            var pNode = this.m_TouchNodeList[i];
            if ( pNode )
            {
                var tRect = this.getRect(pNode);

                if ( cc.Rect.CCRectContainsPoint(tRect,tPoint) )
                {
                    return true;
                }
            }
        }
        return false;
    },
    getRect:function (pNode)
    {
        var rc = cc.RectMake(pNode.getPosition().x , pNode.getPosition().y , pNode.getContentSize().width , pNode.getContentSize().height);

        var scaleY = pNode.getScaleY();
        var scaleX = pNode.getScaleX();

        if ( scaleX != 1.0 )
        {
            rc.size.width *= scaleX;
        }
        if ( scaleY != 1.0 )
        {
            rc.size.height *= scaleY;
        }

        if ( pNode.isIgnoreAnchorPointForPosition() == false )
        {
            var tAnchorPoint = pNode.getAnchorPoint();

            rc.origin.x = ( rc.origin.x - rc.size.width * tAnchorPoint.x ) | 0;
            rc.origin.y = ( rc.origin.y - rc.size.height * tAnchorPoint.y ) | 0;
        }

        return rc;
    },
    getIsTouchForbidden:function()
    {
        return this.m_bTouchForbidden;
    },
    setIsTouchForbidden: function (tEnabled)
    {
        this.m_bTouchForbidden = tEnabled;
    },
    getIsTouchAreaEnabled: function ()
    {
        return this.m_bTouchAreaEnabled;
    },
    setIsTouchAreaEnabled: function (tEnabled)
    {
        this.m_bTouchAreaEnabled = tEnabled;
    }
});

lc.TouchLayer.create = function () {
    var ret = new lc.TouchLayer();
    if (ret && ret.init()) {
        return ret;
    }
    return null;
};